In [1]:
import pandas as pd 
import numpy as np
import sklearn
from sklearn import tree, svm
from sklearn import cluster
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import train_test_split
from sklearn import metrics

In [2]:
cancer_df=pd.read_csv(r'C:\Users\pad39\Downloads\sobar-72.csv')

In [3]:
cancer_df.head(1)

behavior_sexualRisk  behavior_eating  behavior_personalHygine  \
0                   10               13                       12   

   intention_aggregation  intention_commitment  attitude_consistency  \
0                      4                     7                     9   

   attitude_spontaneity  norm_significantPerson  norm_fulfillment  \
0                    10                       1                 8   

   perception_vulnerability  perception_severity  motivation_strength  \
0                         7                    3                   14   

   motivation_willingness  socialSupport_emotionality  \
0                       8                           5   

   socialSupport_appreciation  socialSupport_instrumental  \
0                           7                          12   

   empowerment_knowledge  empowerment_abilities  empowerment_desires  \
0                     12                     11                    8   

   ca_cervix  
0          1

In [4]:
cancer_x, cancer_y = cancer_df.values[:,:19], cancer_df.values[:,19]

In [5]:
score_cancer=[]
for k in range(2,20):
    km = cluster.KMeans(n_clusters=k)
    km.fit(cancer_x)
    score_cancer.append(metrics.silhouette_score(cancer_x,km.labels_))

In [6]:
score_cancer

[0.28012497988495466,
 0.27021840973247135,
 0.2623529362796995,
 0.24856275979868606,
 0.24057734712722498,
 0.2371790164682752,
 0.24643483537330957,
 0.18198398977698974,
 0.19361259082729557,
 0.2013232837786256,
 0.15106595206173976,
 0.19574335631705142,
 0.16493192567974985,
 0.20040788011482263,
 0.2067807950290889,
 0.20916560531999853,
 0.15442538697997815,
 0.16917023359652825]

In [7]:
max(score_cancer)

0.28012497988495466

In [9]:
dt=tree.DecisionTreeClassifier()

In [10]:
params_dt={'criterion':['gini','entropy'],'max_depth':[2,3,4],
          'min_samples_leaf':list(range(2, 20, 2))}
dt_opt=GridSearchCV(dt,params_dt)
dt_opt.fit(cancer_x,cancer_y)

GridSearchCV(estimator=DecisionTreeClassifier(),
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [2, 3, 4],
                         'min_samples_leaf': [2, 4, 6, 8, 10, 12, 14, 16, 18]})

In [11]:
dt_opt.best_params_

{'criterion': 'gini', 'max_depth': 4, 'min_samples_leaf': 2}

In [12]:
dt=tree.DecisionTreeClassifier(max_depth=4, min_samples_leaf=2)
x_train, x_test, y_train, y_test=train_test_split(cancer_x,cancer_y)

In [13]:
dt=dt.fit(x_train,y_train)
dt.score(x_test,y_test)

0.9444444444444444

it would seem the decision tree has won this round now I will compare the tree with svm

In [14]:
svm_clf=svm.SVC(kernel='linear')

In [17]:
param_grid= {'kernel':['linear','rbf', 'poly', 'sigmoid'],'C':[.5,1,2,3,4,5,6,7,8,9,10], 'gamma':['scale','auto']}
svm_opt=GridSearchCV(svm_clf,param_grid)

In [18]:
svm_opt.fit(cancer_x,cancer_y)

GridSearchCV(estimator=SVC(kernel='linear'),
             param_grid={'C': [0.5, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
                         'gamma': ['scale', 'auto'],
                         'kernel': ['linear', 'rbf', 'poly', 'sigmoid']})

In [19]:
svm_opt.best_params_

{'C': 5, 'gamma': 'scale', 'kernel': 'rbf'}

In [20]:
svm_clf2=svm.SVC(kernel='rbf', C=5, gamma='scale')

In [21]:
svm_clf2.fit(x_train,y_train)

SVC(C=5)

In [22]:
svm_clf2.score(x_test, y_test)

0.8333333333333334

After all models were optimized the best model is the decision tree with a score of 94%